### ISSUE TITLE
[bitnami/redis-cluster] CrashLoopBackOff: mkdir: cannot create directory '/opt/bitnami/redis/tmp': Permission denied

### ISSUE THREAD

**Issue by lorantalas** [bitnami/charts#7674]
**Which chart**:
Redis-cluster:15.4.0

**Describe the bug**
I'm trying to install redis-cluster chart and getting a permissions error on creating `'/opt/bitnami/redis/tmp'` directory.

```
helm upgrade --install --wait
        --set usePassword=false
        --set persistence.enabled=false
        --set serviceAccount.name="$KUBE_SERVICE_ACCOUNT"
        --set podAnnotations."app\.gitlab\.com/app"="$CI_PROJECT_PATH_SLUG"
        --set podAnnotations."app\.gitlab\.com/env"="$CI_ENVIRONMENT_SLUG"
        --set tolerations[0].key="node-role.kubernetes.io/test"
        --set tolerations[0].operator="Exists"
        redis bitnami/redis-cluster
```

However containers don't start:

```
NAME                    READY   STATUS             RESTARTS   AGE
redis-redis-cluster-0   0/1     CrashLoopBackOff   25         106m
redis-redis-cluster-1   0/1     CrashLoopBackOff   25         106m
redis-redis-cluster-2   0/1     CrashLoopBackOff   25         106m
redis-redis-cluster-3   0/1     CrashLoopBackOff   25         106m
redis-redis-cluster-4   0/1     CrashLoopBackOff   25         106m
redis-redis-cluster-5   0/1     CrashLoopBackOff   18         71m
```

In the logs there is an error `mkdir: cannot create directory '/opt/bitnami/redis/tmp': Permission denied`

```
redis-cluster 12:25:50.85 Welcome to the Bitnami redis-cluster container
redis-cluster 12:25:50.85 Subscribe to project updates by watching https://github.com/bitnami/bitnami-docker-redis-cluster
redis-cluster 12:25:50.86 Submit issues and feature requests at https://github.com/bitnami/bitnami-docker-redis-cluster/issues
redis-cluster 12:25:50.86 
redis-cluster 12:25:50.86 INFO  ==> ** Starting Redis setup **
redis-cluster 12:25:50.90 WARN  ==> You set the environment variable ALLOW_EMPTY_PASSWORD=yes. For safety reasons, do not use this flag in a production environment.
redis-cluster 12:25:50.91 INFO  ==> Initializing Redis
mkdir: cannot create directory '/opt/bitnami/redis/tmp': Permission denied
```

**Expected behavior**
Create directory /opt/bitnami/redis/tmp without issue. 

**Version of Helm and Kubernetes**:

- Output of `helm version`:

```
version.BuildInfo{Version:"v3.7.0", GitCommit:"eeac83883cb4014fe60267ec6373570374ce770b", GitTreeState:"clean", GoVersion:"go1.16.8"}
```

- Output of `kubectl version`:

```
Client Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.3", GitCommit:"ca643a4d1f7bfe34773c74f79527be4afd95bf39", GitTreeState:"clean", BuildDate:"2021-07-15T21:04:39Z", GoVersion:"go1.16.6", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.5", GitCommit:"aea7bbadd2fc0cd689de94a54e5b7b758869d691", GitTreeState:"clean", BuildDate:"2021-09-15T21:04:16Z", GoVersion:"go1.16.8", Compiler:"gc", Platform:"linux/amd64"}
```

**Comment by miguelaeh** [comment]
Hi @lorantalas ,
Could you check which permissions do that folder have?

Also, have in mind without persistence the cluster will not properly recover nodes.

**Comment by lorantalas** [comment]
Hi, @miguelaeh 
Thank you for your answer! 

```
ls -la /opt/bitnami/redis

total 20
drwxrwxr-x 1 root root 4096 Aug 27 08:39 .
drwxrwxr-x 1 root root 4096 Sep 26 04:31 ..
drwxrwxr-x 1 root root 4096 Aug 27 08:39 bin
drwxrwsrwx 2 root 1001 4096 Oct  4 15:09 etc
drwxrwxr-x 1 root root 4096 Aug 27 08:39 licenses
```

The directory is owned by root but should be owned by 1001.

I tried to resolve the issue using `initContainer`:

```yaml
redis:
  initContainers:
  - name: fix-tmp-dir-permissions
    image: busybox
    command: ['chgrp', '1001', '/opt/bitnami/redis']
    volumeMounts:
    - mountPath: /opt/bitnami/
      name: opt-bitnami
```

However, if I do not set:

```yaml
persistence:
  enabled: true
volumePermissions:
  enabled: true
```

resulting manifest wouldn't contain `initContainer`. 

Because of condition here: https://github.com/bitnami/charts/blob/master/bitnami/redis-cluster/templates/redis-statefulset.yaml#L332:L369

```yaml
{{- $needsVolumePermissions := and .Values.volumePermissions.enabled (and  .Values.persistence.enabled .Values.containerSecurityContext.enabled) }}
{{- if or $needsVolumePermissions .Values.sysctlImage.enabled }}
```

Which I believe is a bug, which could be fixed by adding `.Values.redis.initContainers` to 
if condition on line 333. 

```yaml
{{- if or $needsVolumePermissions (or .Values.redis.initContainers .Values.sysctlImage.enabled) }}
```

If I set those values like this:

```yaml
persistence:
  enabled: true

volumePermissions:
  enabled: true

redis:
  initContainers:
  - name: fix-tmp-dir-permissions
    image: busybox
    command: ['chgrp', '1001', '/opt/bitnami/redis']
    volumeMounts:
    - mountPath: /opt/bitnami/
      name: opt-bitnami
```

Then `initContainer` would be in the resulting manifest for `StatefullSet`:

```yaml
initContainers:
  - name: volume-permissions
    image: docker.io/bitnami/bitnami-shell:10-debian-10-r203
    imagePullPolicy: "Always"
    command: ["/bin/chown", "-R", "1001:1001", "/bitnami/redis/data"]
    securityContext:
      runAsUser: 0
    resources:
      limits: {}
      requests: {}
    volumeMounts:
      - name: redis-data
        mountPath: /bitnami/redis/data
        subPath: 
  - command:
    - chgrp
    - "1001"
    - /opt/bitnami/redis
    image: busybox
    name: fix-tmp-dir-permissions
    volumeMounts:
    - mountPath: /opt/bitnami/
      name: opt-bitnami
```

`volume-permissions` container is run from UID 0, which is not allowed by Pod Security Policy.

```	  
create Pod redis-redis-cluster-0 in StatefulSet redis-redis-cluster failed 
	error: pods "redis-redis-cluster-0" is forbidden: 
		PodSecurityPolicy: 
			unable to admit pod: 
				[spec.initContainers[0].securityContext.runAsUser: 
					Invalid value: 0: running with the root UID is forbidden]
```

**Comment by miguelaeh** [comment]
Hi @lorantalas ,
I guess you are experiencing that issue due to the emptyDir volume that is used when persistence is not enabled. It seems the fsGroup will not be properly set due to the conditions.
Could you create a PR fixing it?

**Comment by lorantalas** [comment]
Hi @miguelaeh, 

I believe that issue with if condition is not a root cause of my issue, neither enabling persistence. 

If I set 

```yaml 
persistence:
  enabled: true
```

Initial error `mkdir: cannot create directory '/opt/bitnami/redis/tmp': Permission denied` would remain. 

With values:

```yaml
persistence:
  enabled: true

redis:
  initContainer:
  - name: fix-tmp-dir-permissions
    image: busybox
    command: ['chgrp', '1001', '/opt/bitnami/redis']
    volumeMounts:
    - mountPath: /opt/bitnami/redis
      name: opt-bitnami-redis
```

Resulting `StatefullSet`, 

```yaml
apiVersion: apps/v1
kind: StatefulSet
metadata:
  name: redis-redis-cluster
  labels:
    app.kubernetes.io/name: redis-cluster
    helm.sh/chart: redis-cluster-6.3.8
    app.kubernetes.io/instance: redis
    app.kubernetes.io/managed-by: Helm
spec:
  selector:
    matchLabels:
      app.kubernetes.io/name: redis-cluster
      app.kubernetes.io/instance: redis
  replicas: 6
  serviceName: redis-redis-cluster-headless
  podManagementPolicy: "Parallel"
  template:
    metadata:
      labels:
        app.kubernetes.io/name: redis-cluster
        helm.sh/chart: redis-cluster-6.3.8
        app.kubernetes.io/instance: redis
        app.kubernetes.io/managed-by: Helm
      annotations:
        checksum/scripts: c1d5bc72dc366b0e6ae707007214f62276fae570ff60ba14716fdaa819b82652
        checksum/secret: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
    spec:
      enableServiceLinks: false
      
      securityContext:
        fsGroup: 1001
        runAsUser: 1001
        sysctls: []
      serviceAccountName: default
      affinity:
        podAffinity:
          
        podAntiAffinity:
          preferredDuringSchedulingIgnoredDuringExecution:
            - podAffinityTerm:
                labelSelector:
                  matchLabels:
                    app.kubernetes.io/name: redis-cluster
                    app.kubernetes.io/instance: redis
                namespaces:
                  - "common"
                topologyKey: kubernetes.io/hostname
              weight: 1
        nodeAffinity:
          
      containers:
        - name: redis-redis-cluster
          image: docker.io/bitnami/redis-cluster:6.2.5-debian-10-r61
          imagePullPolicy: "IfNotPresent"
          securityContext:
            runAsUser: 1001
          command: ['/bin/bash', '-c']
          args:
            - |
              # Backwards compatibility change
              if ! [[ -f /opt/bitnami/redis/etc/redis.conf ]]; then
                  echo COPYING FILE
                  cp  /opt/bitnami/redis/etc/redis-default.conf /opt/bitnami/redis/etc/redis.conf
              fi
              pod_index=($(echo "$POD_NAME" | tr "-" "\n"))
              pod_index="${pod_index[-1]}"
              if [[ "$pod_index" == "0" ]]; then
                export REDIS_CLUSTER_CREATOR="yes"
                export REDIS_CLUSTER_REPLICAS="1"
              fi
              /opt/bitnami/scripts/redis-cluster/entrypoint.sh /opt/bitnami/scripts/redis-cluster/run.sh
          env:
            - name: POD_NAME
              valueFrom:
                fieldRef:
                  fieldPath: metadata.name
            - name: REDIS_NODES
              value: "redis-redis-cluster-0.redis-redis-cluster-headless redis-redis-cluster-1.redis-redis-cluster-headless redis-redis-cluster-2.redis-redis-cluster-headless redis-redis-cluster-3.redis-redis-cluster-headless redis-redis-cluster-4.redis-redis-cluster-headless redis-redis-cluster-5.redis-redis-cluster-headless "
            - name: ALLOW_EMPTY_PASSWORD
              value: "yes"
            - name: REDIS_AOF_ENABLED
              value: "yes"
            - name: REDIS_TLS_ENABLED
              value: "no"
            - name: REDIS_PORT
              value: "6379"
          ports:
            - name: tcp-redis
              containerPort: 6379
            - name: tcp-redis-bus
              containerPort: 16379
          livenessProbe:
            initialDelaySeconds: 5
            periodSeconds: 5
            # One second longer than command timeout should prevent generation of zombie processes.
            timeoutSeconds: 6
            successThreshold: 1
            failureThreshold: 5
            exec:
              command:
                - sh
                - -c
                - /scripts/ping_liveness_local.sh 5
          readinessProbe:
            initialDelaySeconds: 5
            periodSeconds: 5
            # One second longer than command timeout should prevent generation of zombie processes.
            timeoutSeconds: 2
            successThreshold: 1
            failureThreshold: 5
            exec:
              command:
                - sh
                - -c
                - /scripts/ping_readiness_local.sh 1
          resources:
            limits: {}
            requests: {}
          volumeMounts:
            - name: scripts
              mountPath: /scripts
            - name: redis-data
              mountPath: /bitnami/redis/data
              subPath: 
            - name: default-config
              mountPath: /opt/bitnami/redis/etc/redis-default.conf
              subPath: redis-default.conf
            - name: redis-tmp-conf
              mountPath: /opt/bitnami/redis/etc/
      volumes:
        - name: scripts
          configMap:
            name: redis-redis-cluster-scripts
            defaultMode: 0755
        - name: default-config
          configMap:
            name: redis-redis-cluster-default
        - name: redis-tmp-conf
          emptyDir: {}
  volumeClaimTemplates:
    - metadata:
        name: redis-data
      spec:
        accessModes:
          - "ReadWriteOnce"
        resources:
          requests:
            storage: "8Gi"
        
  updateStrategy:
    type: RollingUpdate
```

`securityContext` is there.

However it would be nice to have ability to provide custom `initContainer`. I create PR: https://github.com/bitnami/charts/pull/7701

**Comment by miguelaeh** [comment]
Hi @lorantalas ,
You can provide custom init container using [this](https://github.com/bitnami/charts/blob/cdcf8c1407a9a23b93fadf513be21ca1f9c7c056/bitnami/redis-cluster/values.yaml#L366)
I am confused right now, there should not be any folder mounted at `/opt/bitnami/redis/tmp`.
Can I ask where are you deploying the chart?

**Comment by lorantalas** [comment]
Hi @miguelaeh, 

I could provide `initConintainer` but only in my fork. Because of condition: 

```yaml
{{- $needsVolumePermissions := and .Values.volumePermissions.enabled (and  .Values.persistence.enabled .Values.containerSecurityContext.enabled) }}
{{- if or $needsVolumePermissions .Values.sysctlImage.enabled }}
```

You are right, there is no folder mounted at `/opt/bitnami/redis/tmp` in the chart, however the chart is using image: https://github.com/bitnami/bitnami-docker-redis-cluster

Let's follow scripts in this image:

log: `INFO  ==> ** Starting Redis setup **`
1. [entrypoint.sh](https://github.com/bitnami/bitnami-docker-redis-cluster/blob/a795f1e9cf1e0d3a4b0b85d5674c7431aadbd736/6.2/debian-10/rootfs/opt/bitnami/scripts/redis-cluster/entrypoint.sh
) script runs

log: `INFO  ==> Initializing Redis`
2. [setup.sh](https://github.com/bitnami/bitnami-docker-redis-cluster/blob/a795f1e9cf1e0d3a4b0b85d5674c7431aadbd736/6.2/debian-10/rootfs/opt/bitnami/scripts/redis-cluster/setup.sh
) script runs 
3. which calls `redis_cluster_initialize()` from [librediscluster.sh](https://github.com/bitnami/bitnami-docker-redis-cluster/blob/a795f1e9cf1e0d3a4b0b85d5674c7431aadbd736/6.2/debian-10/rootfs/opt/bitnami/scripts/librediscluster.sh)
4. which calls `redis_configure_default()` from [libredis.sh](https://github.com/bitnami/bitnami-docker-redis-cluster/blob/a795f1e9cf1e0d3a4b0b85d5674c7431aadbd736/6.2/debian-10/rootfs/opt/bitnami/scripts/libredis.sh
)
5. which calls `redis_configure_permissions()` from [libredis.sh](https://github.com/bitnami/bitnami-docker-redis-cluster/blob/a795f1e9cf1e0d3a4b0b85d5674c7431aadbd736/6.2/debian-10/rootfs/opt/bitnami/scripts/libredis.sh)
6. which calls `ensure_dir_exists "${REDIS_BASE_DIR}/tmp"` from `/opt/bitnami/scripts/libfs`
7. which does `mkdir -p "${REDIS_BASE_DIR}/tmp"`

Basically I found only one usage of this folder in the chart itself. In redis config path to PID file is 
`/opt/bitnami/redis/tmp/redis_6379.pid`

https://github.com/bitnami/charts/blob/master/bitnami/redis-cluster/templates/configmap.yaml#L241

I'm deploying the chart to private k8s installation.
